File: ESO.py
import numpy as np
import time
from functools import lru_cache

class ESO:
    def __init__(self, function, pop_size=50, max_iter=1000, max_eval=500000, objective='min', bounds=None, verbose=True):
        self.cache = {}        
        self.ionized_areas_index = [] 
        self.ionized_areas_positions = [] 
        self.storm_power = 0 
        self.bounds = bounds or [(0, 1) for _ in range(2)] 
        self.dim = len(self.bounds) 
        self.function = function 
        self.function_evaluations = 0 
        self.function_evaluations_list = []
        self.lightning = np.array([self.initialize_lightning() for _ in range(pop_size)]) 
        self.n_lightning = pop_size 
        self.objective = objective 
        self.iterations = max_iter 
        self.iteration_current = 0 
        self.field_resistance = 0 
        self.field_intensity = 0 
        self.ke = 0    
        self.verbose = verbose 
        self.stagnation = np.zeros(pop_size, dtype=int) 
        self.objective_values = [] 
        self.iteration_times = []  
        self.best_positions = [] 
        self.history = [[] for _ in range(pop_size)]     
        self.max_evaluations = max_eval    
        self.field_intensity_history = []
        self.field_resistance_history = []
        self.field_elasticity_history = []   
        self.storm_power_history = []       
    
    
    @lru_cache(maxsize=None)
    def exp(self, x):       
        try:           
            x = float(x)            
            # Special cases
            if np.isnan(x):
                return np.nan
            if np.isinf(x):
                return 0.0 if x < 0 else np.inf
                
            # Limit the values to avoid overflow
            if x > 709:  # log(np.finfo(np.float64).max)
                return np.finfo(np.float64).max
            if x < -709:
                return np.finfo(np.float64).min
                
            return np.exp(x)
            
        except Exception as e:
            # In case of unexpected error, return a safe value
            print(f"Warning: exp calculation failed with {str(e)}. Using fallback value.")
            return np.finfo(np.float64).eps 

    @lru_cache(maxsize=None)
    def log(self, x):
       
        try:            
            x = float(x)
            
            # Special cases
            if np.isnan(x):
                return np.nan
            if np.isinf(x):
                return np.inf if x > 0 else np.nan
            if x <= 0:
                # For negative values and zero return a very small value          
                return np.log(np.finfo(np.float64).tiny)
                
            # For values very close to zero
            if x < np.finfo(np.float64).tiny:
                return np.log(np.finfo(np.float64).tiny)
                
            return np.log(x)
            
        except Exception as e:
            # In case of unexpected error, return a safe value
            print(f"Warning: log calculation failed with {str(e)}. Using fallback value.")
            return np.log(np.finfo(np.float64).tiny)
    
    def logistic_curve(self, value, multiplier):
        return 1 / (1 + self.exp(-multiplier * (value - np.abs(self.log(1 - value + 10e-50)))))
          
    def identify_ionized_areas(self):
        self.ionized_areas_index = sorted(range(self.n_lightning), key=lambda i: self.lightning[i]['best_score'], 
                                            reverse=(self.objective == 'max'))[:int(self.n_lightning * (self.field_resistance / 2))]
        self.ionized_areas_positions = [self.lightning[i]['position'] for i in self.ionized_areas_index]
        
    def initialize_lightning(self):                
        if self.ionized_areas_positions:
            # Chooses a random index among the ionized areas
            idx = np.random.randint(len(self.ionized_areas_positions))
            selected_area = self.ionized_areas_positions[idx]
            position = selected_area + self.storm_power 
            position = np.clip(position, *np.array(self.bounds).T)
            
        else:
            bounds_low, bounds_high = np.array(self.bounds).T
            position = np.random.uniform(bounds_low, bounds_high, size=self.dim)
            position = np.clip(position, *np.array(self.bounds).T)
        
        if tuple(position) in self.cache:
            best_score = self.cache[tuple(position)]
        else:
            best_score = self.function(position)
            self.cache[tuple(position)] = best_score
            self.function_evaluations += 1
            self.function_evaluations_list.append(self.function_evaluations)
        
        return {'position': position, 'best_position': np.copy(position), 'best_score': best_score}                   
       
    def calculate_field_resistance(self):        
        positions = np.array([ray['position'] for ray in self.lightning])
        max_diff = np.ptp(positions)  
        max_diff = max(max_diff, 1e-6)  
        self.field_resistance = (np.std(positions) / max_diff)         
   
    def calculate_ke(self):
        beta = self.logistic_curve(self.field_resistance, self.exp(self.field_resistance) / (self.field_resistance + 10e-50))       
        self.ke = self.exp(self.field_resistance) + (self.exp(1 - self.field_resistance) * np.abs(self.log(self.field_resistance + 10e-50))) * beta        
         
    def adjust_field_intensity(self):     
                    
        iteration_factor = self.iteration_current / max(self.iterations, 1)            
        ganma = 1 / (1 + self.exp(-(self.exp(self.field_resistance) / (self.field_resistance + 10e-50)) * 
                                  (self.field_resistance - np.abs(np.log(1 - iteration_factor)))))
        self.field_intensity = 10e-50 + self.ke * ganma       
        
    def calculate_storm_power(self):
        self.storm_power = self.field_resistance * self.field_intensity ** self.ke   
    
    def branch_and_propagate(self, idx):
        # Reinitialize lightning if they have been stagnant              
        if self.stagnation[idx] > 2:
            self.stagnation[idx] = 0
            return self.initialize_lightning()['position']                                                            

        # Propagate the lightning within the high-energy centers
        if idx in self.ionized_areas_index:
            new_position = (self.lightning[idx]['position'] * self.storm_power)           
            # Ensure the position is within the search space boundaries
            new_position = np.clip(new_position, *np.array(self.bounds).T)               
        else:
            # For other lightning, branch them based on the average position of the storm centers
            if not self.ionized_areas_positions:  # Si no hay áreas ionizadas
                # Generar una nueva posición aleatoria dentro de los límites
                bounds_low, bounds_high = np.array(self.bounds).T
                new_position = np.random.uniform(bounds_low, bounds_high, size=self.dim)
            else:
                accumulated_position = np.zeros_like(self.lightning[idx]['position'])
                num_valid_updates = 0
                
                for center_pos in self.ionized_areas_positions:
                    perturbation = np.random.uniform(-self.ke, self.ke, size=self.dim)
                    update = (center_pos + (perturbation * self.storm_power * np.exp(self.ke)))
                    
                    # Verify update
                    if np.all(np.isfinite(update)):
                        accumulated_position += update
                        num_valid_updates += 1
                
                # Calculate mean if uptade was valid
                if num_valid_updates > 0:
                    new_position = accumulated_position / num_valid_updates
                else:
                    # If update not valid generate ramdon position wihtin bounds
                    bounds_low, bounds_high = np.array(self.bounds).T
                    new_position = np.random.uniform(bounds_low, bounds_high, size=self.dim)

            # Ensure the position is within the search space boundaries
            new_position = np.clip(new_position, *np.array(self.bounds).T)

        return new_position
                                     
    def select_path(self, idx, adapted_path):       
        if tuple(adapted_path) in self.cache:
            adapted_score = self.cache[tuple(adapted_path)]
        else:
            adapted_score = self.function(adapted_path)
            self.cache[tuple(adapted_path)] = adapted_score
            self.function_evaluations += 1
            self.function_evaluations_list.append(self.function_evaluations)
        
        if (self.objective == 'min' and adapted_score < self.lightning[idx]['best_score']) or \
            (self.objective == 'max' and adapted_score > self.lightning[idx]['best_score']):
            self.lightning[idx]['position'] = adapted_path
            self.lightning[idx]['best_position'] = adapted_path
            self.lightning[idx]['best_score'] = adapted_score
            self.stagnation[idx] = 0
        else:
            self.stagnation[idx] += 1        
             
    def update_optimal_strike(self, idx, global_best_position, global_best_score):
        if isinstance(global_best_score, np.ndarray):
            global_best_score = global_best_score.item() 

        if (self.objective == 'min' and self.lightning[idx]['best_score'] < global_best_score) or \
        (self.objective == 'max' and self.lightning[idx]['best_score'] > global_best_score):
            global_best_position = self.lightning[idx]['best_position']
            global_best_score = self.lightning[idx]['best_score']
        return global_best_position, global_best_score
    
    def optimize(self):            
        if not callable(self.function):
            raise ValueError("Function must be callable.")
        
        global_best_position = self.lightning[0]['position'].copy()
        global_best_score = self.lightning[0]['best_score']
                    
        for iteration in range(self.iterations):
            if self.function_evaluations >= self.max_evaluations:
                print("Max evaluations reached, stopping optimization.")
                break 
            self.iteration_current = iteration
            iteration_start_time = time.time()
            self.identify_ionized_areas()
            self.adjust_field_intensity()
            self.calculate_field_resistance()
            self.calculate_ke()
            self.calculate_storm_power()
                                        
            for idx in range(self.n_lightning):
                adapted_path = self.branch_and_propagate(idx)
                self.select_path(idx, adapted_path)
                global_best_position, global_best_score = self.update_optimal_strike(idx, global_best_position, global_best_score)
            
            iteration_duration = time.time() - iteration_start_time
            self.iteration_times.append(iteration_duration)
            self.objective_values.append(global_best_score)
            self.best_positions.append(global_best_position)
            
            self.field_intensity_history.append(self.field_intensity)
            self.field_resistance_history.append(self.field_resistance)
            self.field_elasticity_history.append(self.ke)
            self.storm_power_history.append(self.storm_power)

            if self.verbose:
                def format_global_best(global_best):
                    if isinstance(global_best, np.ndarray):
                        global_best = global_best.item()
                    return f"{global_best:.20e}"
                        
                def format_current_best(current_best):
                    if isinstance(current_best, np.ndarray):
                        return np.array2string(current_best)
                    else:
                        return f"{current_best:.20e}"        

                print(f"Iteration: {iteration+1}, Storm P: {self.storm_power }, Field int: {self.field_intensity:.3f}, Field res: {self.field_resistance:.3f}, ke: {self.ke}, Current best: {format_current_best(self.lightning[0]['best_score'])}, Global best: {format_global_best(global_best_score)}, Iteration time: {iteration_duration:.5f} seconds, Func. Eval: {self.function_evaluations}")
                                
        return global_best_position, global_best_score

File: ESO.m
classdef ESO
    properties
        cache
        ionized_areas_index
        ionized_areas_positions
        storm_power
        bounds
        dim
        function_handle
        function_evaluations
        function_evaluations_list
        lightning
        n_lightning
        objective
        iterations
        iteration_current
        field_resistance
        field_intensity
        ke
        verbose
        stagnation
        objective_values
        iteration_times
        best_positions
        history
        max_evaluations
        field_intensity_history
        field_resistance_history
        field_elasticity_history
        storm_power_history
    end
    
    methods
        function obj = ESO(function_handle, pop_size, max_iter, max_eval, objective, bounds, verbose)
            if nargin < 7, verbose = true; end
            if nargin < 6, bounds = [0, 1]; end
            if nargin < 5, objective = 'min'; end
            if nargin < 4, max_eval = 500000; end
            if nargin < 3, max_iter = 1000; end
            if nargin < 2, pop_size = 50; end
            
            obj.cache = containers.Map();
            obj.ionized_areas_index = [];
            obj.ionized_areas_positions = [];
            obj.storm_power = 0;
            obj.bounds = bounds;
            obj.dim = size(bounds, 1);
            obj.function_handle = function_handle;
            obj.function_evaluations = 0;
            obj.function_evaluations_list = [];
            obj.lightning = arrayfun(@(~) obj.initialize_lightning(), 1:pop_size, 'UniformOutput', false);
            obj.n_lightning = pop_size;
            obj.objective = objective;
            obj.iterations = max_iter;
            obj.iteration_current = 0;
            obj.field_resistance = 0;
            obj.field_intensity = 0;
            obj.ke = 0;
            obj.verbose = verbose;
            obj.stagnation = zeros(pop_size, 1);
            obj.objective_values = [];
            obj.iteration_times = [];
            obj.best_positions = [];
            obj.history = cell(pop_size, 1);
            obj.max_evaluations = max_eval;
            obj.field_intensity_history = [];
            obj.field_resistance_history = [];
            obj.field_elasticity_history = [];
            obj.storm_power_history = [];
        end
        
        function value = exp(obj, x)
            if x > 10e1
                value = inf;
            elseif x < -10e1
                value = 0.0;
            else
                value = exp(x);
            end
        end
        
        function value = log(obj, x)
            value = log(x);
        end
        
        function obj = identify_ionized_areas(obj)
            [~, sorted_indices] = sort(cellfun(@(light) light.best_score, obj.lightning), 'descend');
            if strcmp(obj.objective, 'min')
                sorted_indices = flip(sorted_indices);
            end
            num_ionized = ceil(obj.n_lightning * (obj.field_resistance / 2));
            obj.ionized_areas_index = sorted_indices(1:num_ionized);
            obj.ionized_areas_positions = cellfun(@(idx) obj.lightning{idx}.position, num2cell(obj.ionized_areas_index), 'UniformOutput', false);
        end
        
        function lightning = initialize_lightning(obj)
            if ~isempty(obj.ionized_areas_positions)
                idx = randi(length(obj.ionized_areas_positions));
                selected_area = obj.ionized_areas_positions{idx};
                position = selected_area + obj.storm_power;
                position = min(max(position, obj.bounds(:, 1)'), obj.bounds(:, 2)');
            else
                position = (obj.bounds(:, 2)' - obj.bounds(:, 1)') .* rand(1, obj.dim) + obj.bounds(:, 1)';
            end
            
            pos_key = mat2str(position);
            if isKey(obj.cache, pos_key)
                best_score = obj.cache(pos_key);
            else
                best_score = obj.function_handle(position);
                obj.cache(pos_key) = best_score;
                obj.function_evaluations = obj.function_evaluations + 1;
                obj.function_evaluations_list = [obj.function_evaluations_list, obj.function_evaluations];
            end
            
            lightning.position = position;
            lightning.best_position = position;
            lightning.best_score = best_score;
        end
        
        function obj = calculate_field_resistance(obj)
            positions = cell2mat(cellfun(@(ray) ray.position, obj.lightning, 'UniformOutput', false)');
            max_diff = max(range(positions), 1e-50);
            obj.field_resistance = std(positions) / max_diff;
        end
        
        function obj = calculate_ke(obj)
            beta = 1 / (1 + obj.exp(-(obj.exp(obj.field_resistance) / (obj.field_resistance + 10e-50)) * ...
                (obj.field_resistance - abs(obj.log(1 - obj.field_resistance)))));
            obj.ke = obj.exp(obj.field_resistance) + (obj.exp(1 - obj.field_resistance) * abs(obj.log(obj.field_resistance + 10e-50))) * beta;
        end
        
        function obj = adjust_field_intensity(obj)
            iteration_factor = obj.iteration_current / max(obj.iterations, 1);
            gamma = 1 / (1 + obj.exp(-(obj.exp(obj.field_resistance) / (obj.field_resistance + 10e-50)) * ...
                (obj.field_resistance - abs(obj.log(1 - iteration_factor)))));
            obj.field_intensity = 10e-50 + obj.ke * gamma;
        end
        
        function obj = calculate_storm_power(obj)
            obj.storm_power = obj.field_resistance * obj.field_intensity ^ obj.ke;
        end
        
        function new_position = branch_and_propagate(obj, idx)
            if obj.stagnation(idx) > round(obj.ke)
                obj.stagnation(idx) = 0;
                new_position = obj.initialize_lightning().position;
                return;
            end
            
            if any(idx == obj.ionized_areas_index)
                new_position = obj.lightning{idx}.position * obj.storm_power;
                new_position = min(max(new_position, obj.bounds(:, 1)'), obj.bounds(:, 2)');
            else
                if ~isempty(obj.ionized_areas_positions)
                    new_position = zeros(1, obj.dim);
                    for i = 1:length(obj.ionized_areas_positions)
                        perturbation = (2 * rand(1, obj.dim) - 1) * obj.ke;
                        new_position = new_position + (obj.ionized_areas_positions{i} + (perturbation * obj.storm_power * obj.exp(obj.ke)));
                    end
                    new_position = new_position / length(obj.ionized_areas_positions);
                else
                    new_position = obj.initialize_lightning().position;
                end
                new_position = min(max(new_position, obj.bounds(:, 1)'), obj.bounds(:, 2)');
            end
        end
        
        function obj = select_path(obj, idx, adapted_path)
            pos_key = mat2str(adapted_path);
            if isKey(obj.cache, pos_key)
                adapted_score = obj.cache(pos_key);
            else
                adapted_score = obj.function_handle(adapted_path);
                obj.cache(pos_key) = adapted_score;
                obj.function_evaluations = obj.function_evaluations + 1;
                obj.function_evaluations_list = [obj.function_evaluations_list, obj.function_evaluations];
            end
            
            if (strcmp(obj.objective, 'min') && adapted_score < obj.lightning{idx}.best_score) || ...
               (strcmp(obj.objective, 'max') && adapted_score > obj.lightning{idx}.best_score)
                obj.lightning{idx}.position = adapted_path;
                obj.lightning{idx}.best_position = adapted_path;
                obj.lightning{idx}.best_score = adapted_score;
                obj.stagnation(idx) = 0;
            else
                obj.stagnation(idx) = obj.stagnation(idx) + 1;
            end
        end
        
        function [global_best_position, global_best_score] = update_optimal_strike(obj, idx, global_best_position, global_best_score)
            if (strcmp(obj.objective, 'min') && obj.lightning{idx}.best_score < global_best_score) || ...
               (strcmp(obj.objective, 'max') && obj.lightning{idx}.best_score > global_best_score)
                global_best_position = obj.lightning{idx}.best_position;
                global_best_score = obj.lightning{idx}.best_score;
            end
        end
        
        function [global_best_position, global_best_score] = optimize(obj)
            if ~isa(obj.function_handle, 'function_handle')
                error('Function must be callable.');
            end
            
            global_best_position = obj.lightning{1}.position;
            global_best_score = obj.lightning{1}.best_score;
            
            for iteration = 1:obj.iterations
                if obj.function_evaluations >= obj.max_evaluations
                    disp('Max evaluations reached, stopping optimization.');
                    break;
                end
                
                obj.iteration_current = iteration;
                iteration_start_time = tic;
                obj = obj.identify_ionized_areas();
                obj = obj.adjust_field_intensity();
                obj = obj.calculate_field_resistance();
                obj = obj.calculate_ke();
                obj = obj.calculate_storm_power();
                
                for idx = 1:obj.n_lightning
                    adapted_path = obj.branch_and_propagate(idx);
                    obj = obj.select_path(idx, adapted_path);
                    [global_best_position, global_best_score] = obj.update_optimal_strike(idx, global_best_position, global_best_score);
                end
                
                iteration_duration = toc(iteration_start_time);
                obj.iteration_times = [obj.iteration_times, iteration_duration];
                obj.objective_values = [obj.objective_values, global_best_score];
                obj.best_positions = [obj.best_positions; global_best_position];
                obj.field_intensity_history = [obj.field_intensity_history, obj.field_intensity];
                obj.field_resistance_history = [obj.field_resistance_history, obj.field_resistance];
                obj.field_elasticity_history = [obj.field_elasticity_history, obj.ke];
                obj.storm_power_history = [obj.storm_power_history, obj.storm_power];
                
                if obj.verbose
                    fprintf('Iteration: %d, Storm Power: %.5f, Field Intensity: %.3f, Field Resistance: %.3f, ke: %.3f, Current Best: %.20e, Global Best: %.20e, Iteration Time: %.5f seconds, Func. Eval: %d\n', ...
                        iteration, obj.storm_power, obj.field_intensity, obj.field_resistance, obj.ke, obj.lightning{1}.best_score, global_best_score, iteration_duration, obj.function_evaluations);
                end
            end
        end
    end
end
